File: service/service-samples/sample-service-rest-oracle-book/src/main/java/io/americanexpress/service/book/rest/service/CreateBookService.java
/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package io.americanexpress.service.book.rest.service;

import io.americanexpress.data.oracle.book.dao.BookRepository;
import io.americanexpress.data.oracle.book.entity.BookEntity;
import io.americanexpress.service.book.rest.model.CreateBookRequest;
import io.americanexpress.service.book.rest.model.CreateBookResponse;
import io.americanexpress.service.book.rest.service.helper.BookServiceMapper;
import io.americanexpress.synapse.framework.exception.ApplicationClientException;
import io.americanexpress.synapse.framework.exception.model.ErrorCode;
import io.americanexpress.synapse.service.rest.service.BaseCreateService;
import org.springframework.http.HttpHeaders;
import org.springframework.stereotype.Service;

/**
 * {@code CreateBookService} Creates a new book resource.
 */
@Service
public class CreateBookService extends BaseCreateService<CreateBookRequest, CreateBookResponse> {

    /**
     * bookRepository
     */
    private final BookRepository bookRepository;

    public CreateBookService(BookRepository bookRepository) {
        this.bookRepository = bookRepository;
    }

    /**
     * executeCreate will be used to create a book resource by request.
     * @param request the create book request
     * @return a create book response
     */
    @Override
    protected CreateBookResponse executeCreate(HttpHeaders headers, CreateBookRequest request) {
        if (isTitlePresent(request.getTitle())) {
            BookEntity bookEntity = bookRepository.save(BookServiceMapper.populateBookEntityForCreation(request));
            return BookServiceMapper.populateCreateBookResponse(bookEntity);
        } else {
            throw new ApplicationClientException("Bad request", ErrorCode.GENERIC_4XX_ERROR, (String[]) null);
        }
    }

    private boolean isTitlePresent(String title) {
        return bookRepository.findByTitle(title) == null;
    }
}

File: service/service-samples/sample-service-kotlin/src/main/kotlin/io/americanexpress/synapse/rest/service/CreateBookService.kt
/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package io.americanexpress.synapse.rest.service

import io.americanexpress.synapse.rest.model.BookRequest
import io.americanexpress.synapse.rest.model.BookResponse
import io.americanexpress.synapse.service.rest.service.BaseCreateService
import org.springframework.http.HttpHeaders
import org.springframework.stereotype.Service
import java.util.*

@Service
class CreateBookService : BaseCreateService<BookRequest, BookResponse>() {

    override fun executeCreate(headers: HttpHeaders?, request: BookRequest?): BookResponse {
        val actualRequest = request ?: BookRequest(UUID.randomUUID().toString(), "Synapse", "Gabriel")
        return BookResponse(title = actualRequest.title, author = actualRequest.author)
    }
}